### PR TITLE
Proc macro server: Load `ProcMacroHost`

### PR DESCRIPTION
**Pull request by Draggu**
**Stack**:
- #1686
- #1685
- #1684
- #1683
- #1718
- #1682
- #1681
- #1680 ⬅

⚠️ *Part of a stack created by [spr](https://github.com/ejoffe/spr). Do not merge manually using the UI - doing so may have unexpected results.*

### CODE REVIEW

File: scarb/src/bin/scarb/commands/proc_macro_server.rs [hunk 1]
```diff
@@ -1,7 +1,52 @@
...
+        &FeaturesOpts {
+            features: FeaturesSelector::AllFeatures,
+            no_default_features: false,
+        },
+        &ws,
+    )?;
+
+    for unit in &compilation_units {
+        if let CompilationUnit::ProcMacro(_) = unit {
+            ops::compile_unit(unit.clone(), &ws)?;
```
Comment (mkaput) [inline scarb/src/bin/scarb/commands/proc_macro_server.rs:23]: why are you compiling everything here?

**Review by mkaput — CHANGES_REQUESTED** [review]

File: scarb/src/bin/scarb/commands/proc_macro_server.rs [hunk 2]
```diff
@@ -1,7 +1,52 @@
...
+    compiler::{plugin::proc_macro::ProcMacroHost, CairoCompilationUnit, CompilationUnit},
+    core::{Config, Workspace},
+    ops::{self, FeaturesOpts, FeaturesSelector},
+};
 
 #[tracing::instrument(skip_all, level = "info")]
-pub fn run(_config: &Config) -> Result<()> {
-    unimplemented!()
+pub fn run(config: &mut Config) -> Result<()> {
+    let ws = ops::read_workspace(config.manifest_path(), config)?;
```
Comment (Arcticae) [inline scarb/src/bin/scarb/commands/proc_macro_server.rs:10]: What if this changes in time (i.e. more crates are added)

Comment (mkaput) [reply scarb/src/bin/scarb/commands/proc_macro_server.rs:10]: We will detect this on a client (LS) and just restart the server

File: scarb/src/bin/scarb/commands/proc_macro_server.rs (hunk 1 above)
Comment (Draggu) [reply scarb/src/bin/scarb/commands/proc_macro_server.rs:23]: Not sure what could be omitted. If you have better inside can you type here what could be removed?

Comment (mkaput) [reply scarb/src/bin/scarb/commands/proc_macro_server.rs:23]: Compiling creates the `RootDatabase`, runs full compilation and outputs artefacts. That's nothing we can make use of, so we are safe to skip this.

The rest of the code here all looks okay for me here. We basically need to `scarb fetch` before booting up, so you just need to:
1. Read the workspace
2. Resolve it, including downloading all dependencies.
3. Generate compilation units to gather all compiler db parameters.
4. Create proc macro host.

Comment (maciektr) [reply scarb/src/bin/scarb/commands/proc_macro_server.rs:23]: @mkaput but those are only macros -> per match one line prior

Comment (mkaput) [reply scarb/src/bin/scarb/commands/proc_macro_server.rs:23]: > @mkaput but those are only macros -> per match one line prior

I don't understand, could you elaborate please?

Comment (maciektr) [reply scarb/src/bin/scarb/commands/proc_macro_server.rs:23]:
```
        if let CompilationUnit::ProcMacro(_) = unit {
            ops::compile_unit(unit.clone(), &ws)?;
```
Unit is only compiled, if it fulfills the `let CompilationUnit::ProcMacro(_) = unit` match, thus is guaranteed to represent a procedural macro.

Comment (mkaput) [reply scarb/src/bin/scarb/commands/proc_macro_server.rs:23]: 🤦🏻 I didn't notice this. @Draggu Could you please add a comment elaborating that we are compiling procedural macros here?

**Review by mkaput — APPROVED** [review]

**Review by piotmag769 — APPROVED** [review]

**Pull request merged**